### ISSUE TITLE
Complex data relationships using fixtures and/or factories

### ISSUE THREAD

**Issue by pfgilbert** [samselikoff/ember-cli-mirage#443]
I've been unable to figure out how to support the following data model with Mirage fixtures or factories:

Two models: Vehicles and Cars, where Vehicles hasMany Cars.

Let's say I've got two Vehicle records, western_vehicles and eastern_vehicles, and just one Car record, honda.  Both western_vehicles and eastern_vehicles need to include honda, so the embedded IDs should look something like this for both western_vehicles and eastern_vehicles:
  vehicle_ids: [id_of_honda]

But it's not obvious to me how to do that.  It seems that child objects, the "honda" record in this case, are required to have the ID of their parent, but honda has two parents.  

I've tried this with fixtures, and if I simply put the list of ids in the parent's fixture file, and leave the parent ID out of the child records, the child objects don't get created. I haven't tried to do something similar with factories, but it seems likely that I'll have the same issue, since I seem to be required to give the parent ID when I create the objects:

server.create('car', {name: honda, vehicle_id: <some parent ID>});

I'm using the current beta release, by the way, to take advantage of the ORM and serializers (nice!), in case that's a factor.

Am I missing something?

**Comment by samselikoff** [comment]
Sorry you're running into trouble! I think it'd be easiest for me to assist if I can see what an expected response looks like. Could you write out an example of a response your Ember app expects?

**Comment by pfgilbert** [comment]
Thanks.  Here's a simplified version of the JSON that exhibits the problem:

```{
  "top_object": {
    "id": 1,
    "name": "A Name",
    "vehicle_ids": [
      "eastern",
      "western"
    ],
  },
  "cars": [
    {
      "id": 1405,
      "name": "Honda Civic"
      // "vehicle_id": ???
    },
    {
      "id": 1438,
      "name": "Ford Taurus"
      // "vehicle_id": ???
    },
    {
      "id": 1431,
      "name": "Smart Car"
      // "vehicle_id": ???
    },

  ],
  "trucks": [
    {
      "id": 11,
      "name": "Ford Bronco"
      // "vehicle_id": ???
    },
    {
      "id": 12,
      "name": "Kamaz"
      // "vehicle_id": ???
    }
  ],
  "vehicles": [
    {
      "id": "eastern",
      topObjectId: "1", // Have to add this or the vehicle doesn't get created
      "car_ids": [
          1405,
          1438
      ],
      "truck_ids": [
        11,
        12
      ]
    },
    {
      "id": "western",
      topObjectId: "1", // Have to add this or the vehicle doesn't get created
      "car_ids": [
          1405,
          1438,
          1431
      ],
      "truck_ids": [
        11
      ]
    },
  ],
}```

Although the actual JSON my Ember app ingests doesn't have those `topObjectId` values in it, if I don't add them to the Mirage fixtures, the vehicle objects don't get created, even though there's a list of `vehicle_ids` at the top level.  Then, the car and truck objects are not being created because I can't put  vehicle IDs into their fixtures, since many of them have more than one parent vehicle.  

The fixtures seem to require the foreign keys to be on both sides of the relationship.

As I said, I haven't tried this with factories yet, but it seems likely I'll have the same problem.  How can I create a single "car" record in the DB that two vehicle records reference?  

Thanks for your help.  I'm hoping that there's just something I'm not understanding or have set up incorrectly.

**Comment by pfgilbert** [comment]
By the way, are you aware that the fixtures and testing links in the v2 documentation are broken?  

http://www.ember-cli-mirage.com/docs/v0.2.x/fixtures/
http://www.ember-cli-mirage.com/docs/v0.2.x/testing/

I realize this isn't the release version, so maybe they're still a work in progress.

**Comment by pfgilbert** [comment]
Looking at the mirage code today, it does look like there's an assumption that the foreign key will be in the attrs of the child object, which would prevent the kind of many-to-many relationship I'm trying to create.

**Comment by pfgilbert** [comment]
I looked at the JSON response generated by Mirage, and instead of returning:

```
      "car_ids": [
          1405,
          1438
      ],
      "truck_ids": [
        11,
        12
      ]
```
above, those lists would be empty:

```
      "car_ids": [
      ],
      "truck_ids": [
      ]
```

So, I think this is an issue with the serialization, but not a simple one, because as I mentioned above, it looks like the corresponding foreign keys have to exist in the child objects, and that doesn't allow for a many-to-many relationship.

**Comment by samselikoff** [comment]
Ok, there's a lot going on here. Typically when I model many-to-many relationships, my backend has a join model, which has to `has_many` relationships. So that's one path we could take.

In the response above, it looks like you added some data to try to accommodate Mirage's ORM. Could you just post exactly what your Ember app expects - exactly what a typical server response would look like - so I can see what the best way to mock this out in Mirage today is? Then, if we identify some missing pieces (e.g. m-2-m support in the ORM), I can add + prioritize those issues for development.

Thanks!

**Comment by pfgilbert** [comment]
Sure.  Here's the GET request: `/api/top_objects/1`

And the response I would expect:

```
{
  "top_object": {
    "id": 1,
    "name": "A Name",
    "vehicle_ids": [
      "eastern",
      "western"
    ],
  },
  "cars": [
    {
      "id": 1405,
      "name": "Honda Civic"
    },
    {
      "id": 1438,
      "name": "Ford Taurus"
    },
    {
      "id": 1431,
      "name": "Smart Car"
    },
  ],
  "trucks": [
    {
      "id": 11,
      "name": "Ford Bronco"
    },
    {
      "id": 12,
      "name": "Kamaz"
    }
  ],
  "vehicles": [
    {
      "id": "eastern",
      "car_ids": [
          1405,
          1438
      ],
      "truck_ids": [
        11,
        12
      ]
    },
    {
      "id": "western",
      "car_ids": [
          1405,
          1438,
          1431
      ],
      "truck_ids": [
        11
      ]
    },
  ],
}
```

**Comment by samselikoff** [comment]
Ok, thanks. And what does the data model look like on your actual server? Also, what language/framework is the server written in? Finally, what's the data model in the Ember app?

**Comment by pfgilbert** [comment]
Thanks for your patience on this.

The backend is a Ruby/Rails app.  We're using serializers to produce the JSON response from the model, with `has_many` relationships and `embed_ids`, but no belongs_to, so my top_object serializer has:
   ` has_many :vehicles, embed: :ids`

and vehicles has:
   ` has_many :cars, embed: :ids`
   ` has_many :trucks, embed: :ids`

In the Ember app, the top-object model has:
   'vehicles: DS.hasMany('vehicle')`

The vehicle object has:
   `cars: DS.hasMany('car')`
   `trucks: DS.hasMany('truck')`

The car and truck models in the Ember app have no relationships defined.

This is a simplified (and obfuscated) version of our data model, to just focus on the problem relationships, but I think it correctly represents the issue.

Ember seems to be able to rely on the lists of embedded IDs for loading child objects, where the Mirage serializer is expecting the child objects to contain the ID of their parent.

**Comment by pfgilbert** [comment]
Actually, it's not the serializer.  It's the `has_many`:

```
      get: function() {
        var models = association._cachedChildren || [];

        if (!this.isNew()) {
          var query = {[foreignKey]: this.id};
          var savedModels = schema[association.target].where(query);

          models = savedModels.mergeCollection(models);
        }

        return models.map(model => model.id);
      },
```

If the child object doesn't contain the parent id its model object isn't returned.  The calling code relies on that list of child objects rather than on its list of embedded child IDs.

**Comment by pfgilbert** [comment]
This may be much simpler than I thought.  The docs for Serializers say:

```
embed

default false

Set whether related models should be embedded or sideloaded.

By default this false, so relationships are sideloaded.
```

But, this code in serializer-registry.js hardcodes embedRelatedIds to `true`:

```
  _serializeSideloadedModelResponse(model, topLevelIsArray = false, allAttrs = {}, root = null) {
    let serializer = this._serializerFor(model);

    // Add this model's attrs
    this._augmentAlreadySerialized(model);
    let modelAttrs = this._attrsForModel(model, false, true);
```
Isn't that a mistake?  If I change it to `false`, everything works the way I would expect it to.

**Comment by pfgilbert** [comment]
I created a PR for that change: https://github.com/samselikoff/ember-cli-mirage/pull/446

**Comment by samselikoff** [comment]
Interesting. I think I've only ever worked on rails apps where the foreign key is included in the serializer, e.g.

```rb
# serializers/vehicle_serializer.rb
class VehicleSerializer < ApplicationSerializer
  attributes :id, :top_object_id

  has_many :cars, embed: :ids
  has_many :trucks, embed: :ids
end
```

You can still have the `embed :ids` set for the `top_object`, but this way whenever a `vehicle` comes through its fk is present. Even if not defined on the Ember Data side, I thought this was done so if you ever, say, PUT a modification to this record and change its owner, the serializer would confirm by reporting the new fk.

I didn't realize sometimes people use the `has_many` side *only* to declare relationships. Originally I used to see people use both, but when the single-source-of-truth branch of Ember Data came out, all you needed was to include the foreign key on the child models, and ED would work out all relationships from there (include hasManys).

In your rails models, you do have fks on children, correct? That is, your `car` and `truck` model each have a `vehicle_id` attribute?

I'll need to work on this a bit more to come up with a solution, but all of this being said I don't see why the child fks should actually cause any problems. Still, if this is what your actual API responds with (and presumably how others also develop their APIs), we should support this in Mirage. It should only require some changes in the Serializer layer.

**Comment by pfgilbert** [comment]
In general, in our Rails models, we are storing FKs on children, yet, but we aren't necessarily serializing the objects for the API calls exactly the way they're organized in the model.  And, I didn't realize this until you asked, but in our Rails app, the equivalent of the Vehicles model isn't actually an ActiveRecord model.  It's just a class that's created on the fly and used for the serialization.

I appreciate your taking time to think about this.  I realize it may not be the most common case, but since Ember and Ember Data support it, we're probably not the only ones taking advantage of it.

As for my PR, it causes some tests to fail.  I haven't had a chance to investigate that yet.  Let me know if you think I should, or if you have a different fix in mind.

In general, though, aren't you considering embed and sideload to be mutually exclusive, so when objects are sideloaded, embedRelatedIDs should be false?

**Comment by samselikoff** [comment]
Yep, I totally agree, since Ember Data supports it it should be in Mirage.

Ok so, back to the particulars. if I use the base serializer in Mirage, `embed` defaults to `false`, and as you say this makes `embedRelatedIDs` true. But isn't that what you want?

For instance in a simple app, I have a `post` that `hasMany` `comments`. If I have the following as my `post` serializer:

```js
import Serializer from 'ember-cli-mirage/serializers/active-model-serializer';

export default Serializer.extend({

  relationships: ['comments']

});
```

then the response to GET `/posts/1` is

```js
post: {
  id: 1,
  title: 'Lorem ipsum',
  date: "2015-12-22T13:59:45.175Z",
  comment_ids: [1, 2, 3]
},
comments: [
  ...
]
```

which is what you want, correct? The `has_many` relationship has the array of ids. Isn't that what your after?

Btw the foreign key will also show up in each serialized comment, but you can use the `attributes` key to whitelist which attrs you want to show up:

```js
// serializers/comment.js
import Serializer from 'ember-cli-mirage/serializers/active-model-serializer';

export default Serializer.extend({

  attributes: ['id', 'text'] // ignore postId by not specifying it here

});
```

Does this accomplish what you're after?

**Comment by samselikoff** [comment]
I think I'm understanding now (sorry it took me so long), above you state you had to add the fk to the fixtures even though it's not in the response set. That's fine. Fixtures (or factories) are about setting up your Mirage database, but you can still customize the *serializer* so that those fks are not returned. Does that make sense? Mirage needs to know how to find the related models, and it uses fks to do that.

**Comment by pfgilbert** [comment]
I'm using the ActiveRecordSerializer, as you did in your example above.  Yes, the response you show above is what I want, and it's what I'll get if the child objects in the fixture data contain FKs.  If they don't contain FKs, and I force `embedRelatedIds` to false, the serializer code drops through to pick up the listed of IDs from the `attrs`, which works perfectly.  So if my fixture data has lists of child IDs in each parent, setting `embedRelatedIds` to false gives me the correct behavior.

Above, where I pasted in a section from your docs, I think I was confusing `embed` with `embedRelatedIds`.  You're right that what I want is side-loaded AND `embedRelatedIds`, which means my PR isn't doing the right thing.  It just happens that in this particular case, turning off `embedRelatedIds` fixes my problem, since my fixture data has the lists of IDs.  Unfortunately, I don't think there's anything I can override in the serialization, currently, to get the behavior I want, though.  The code inside `if (embedRelatedIds)` (which calls the has_many code) will have to be changed before Mirage will work with my data model.

Thanks again for all the time you've taken to understand this.

**Comment by samselikoff** [comment]
Ok, I think I see. So originally, you were adding the list of child IDs to your fixtures? I imagine you did this because this is how your API responses are formatted? If so, this is a communication error on Mirage's part. Fixtures are used to seed Mirage's db, and those fixture files should be in the format Mirage expects - regardless of your particular API format. This is because Mirage's ORM expects to find foreign keys on child models. It's independent of the particular serialization format of your API.

Perhaps this is something I need to clear up in the documentation. Do you think that was the main issue here? And, have you found a good working solution for your data model now? One last question - are you writing the fixture file from hand, or copying the API response from your backend?

**Comment by pfgilbert** [comment]
The main issue, which I still have no solution for, is that Mirage expects each child object to contain the FK of its parent, which, unless I'm missing something, makes it impossible to create a many-to-many relationship.  If Mirage were changed to rely on the list of child IDs each parent provides (see the car_ids and truck_ids above), if the parent provides one, that would solve the problem.  The PR I created did solve the issue in my particular case, in case you want to look at that. I've since closed it, because it was incorrect in other ways.

I don't think I'll be able to use Mirage until this is solved.  I've been unable to come up with a way to get around it.  If you think there's some way to support a many-to-many relationship, let me know.

To create the fixture files I'm copying our API response into the files and then adding some FKs to them where Mirage expects them and we're not generating them. That fix-up isn't really an issue.  I can live with that, if I can solve the lack of support for many-to-many relationships.

**Comment by pfgilbert** [comment]
I just reread this, above:

> when the single-source-of-truth branch of Ember Data came out, all you needed was to include the foreign key on the child models, and ED would work out all relationships from there (including hasManys).

I'm not sure how that would work, unless you're allowed to include a list of FKs on the child, not just a single FK.  I'll take a look at how this is done in Ember Data.

**Comment by pfgilbert** [comment]
Okay, so this is how ED is doing it:

> At a high-level, the solution is this: instead of internally storing the relationships as state on each record that needs to be synchronized, we maintain a single entity in memory that represents the logical relationship between the records. That entity exists regardless of which side of the relationship the application has loaded.

That's a more complicated solution than Mirage needs, I think, because ED will do the work for you if you let it.  That's why the change in my PR worked (for me...if not for your tests): it allowed the lists of IDs each parent object had to be passed through to ED, and ED didn't care that there were no corresponding IDs on the child side.

**Comment by samselikoff** [comment]
The way I've always modeled many-to-many relationships is with a join model, and I've always modeled this join model in Ember Data as well as in Rails.

For example if there's a m2m relationship between `posts` and `tags`, there'd be a `post_tag` join model, and the models would look something like this:

```rb
# models/post.rb
class Post < ActiveRecord::Base
  has_many :post_tags
  has_many :tags, through: :post_tags
end

# models/tag.rb
class Tag < ActiveRecord::Base
  has_many :post_tags
  has_many :posts, through: :post_tags
end

# models/post_tag.rb
class PostTag < ActiveRecord::Base
  belongs_to :post
  belongs_to :tag
end
```

I believe you don't need to define the join model explicitly, but even still the join table exists.

Then, I model all three classes in Ember Data as well. This way, all modeling works through `hasMany` and `belongsTo`. If you need to delete a join, you send a DELETE request to `/post_tags/:id`, and this would drop a post-tag relationship. I believe this is the most conventional way to do things, and when I used to do it differently I found myself framework-fighting, either on the Rails side or the Ember side.

Since Mirage is a mock of your backend, if the join table exists in your backend - and my working assumption is that is does - it should probably exist in Mirage. If it turns out that people are regularly doing m2m without separate join models, then we can talk about adding that. But I believe the Rails guides, DHH's talks etc. typically present m2m via a join table. I of course could be wrong.

What do you think? Does your m2m use a join table?

**Comment by pfgilbert** [comment]
No, we're not using a join table for this (for what would be the vehicle table in my example).  I understand that that would fix the problem, and we may have to consider doing that to support Mirage, but since we don't need it, or haven't yet, for any other reason, it might be difficult for me to sell that change.

I just want to make clear that if I doctor the Mirage code a bit to not try to calculate the set of child IDs itself and instead just go with the list that I send in with the parent's fixture data, Mirage and Ember/ED do the right thing.  It seems like it wouldn't take a big change to have Mirage pay attention to those lists of IDs, but I haven't been able to put any more time into a better pull request for that change.

**Comment by samselikoff** [comment]
You should absolutely not have to restructure your data model to support Mirage, assuming your data model is not something totally outlandish (which it doesn't seem like it is).

I'm confused, I assumed there was always a join table in every m2m relationship, even if the join record was transparent to the serializer layer (i.e. even if you serialized the relationship using `child_ids: []`.

If your backend doesn't have a join table, how are you storing the relationship? Does your `vehicle` table have a `car_ids` array column or something?

**Comment by pfgilbert** [comment]
On the Rails end of things, the `vehicle` model isn't an ActiveRecord one.  It's just there to organize some things in memory, and for the serialization.

**Comment by samselikoff** [comment]
Ahh, I see. Ok. Thanks for taking the time to explain all this.

In Rails (and servers in general), m2m relationships use a join table. Even when you use `has_and_belongs_to_many` in Rails to avoid defining an AR model for the join table, the join table itself must still be created, to store the relationship. This is why I modeled m2m relationships like this in Mirage - the assumption being, the easiest way to mock out a typical m2m relationship on your server would be to replicate it in Mirage (using the three tables).

I do understand you can fudge the response to get what you're after, but I'm more interested if there's an underlying architectural change that needs to be made to Mirage to accommodate these sorts of situations in the future. Discussing this is helping me understand your use case.

Curious, how do you generate the list of ids in memory? What data is used to determine the relationship?

**Comment by pfgilbert** [comment]
Sorry for the delay.  The serializers on the Rails side are using `:embed_ids`.  I'll send you a more complete explanation soon.

**Comment by samselikoff** [comment]
Right, the `:embed_ids` bit I understand. But if you're not using a join table to store the vehicle associations, where are you storing them?

For example, the vehicle with id `eastern` has `car_ids` of `[1405, 1438]`. Where are those two ids being stored in your actual backend? And if they're not being persisted in the db, how do you derive them?

**Comment by cs3b** [comment]
I've another example that might add something to this discussion.
I was unsure if I should create a new issue, if yes just let me know.

Let's say we have in Ember.js this kind of situation:

````javascript
// app/user/model.js
import Model from 'ember-data/model';
import attr from 'ember-data/attr';
import { hasMany } from 'ember-data/relationships';

import Ember from 'ember';

export default Model.extend({
  username: attr('string'),
  // (...)

  followees: hasMany('user', {
    inverse: 'followers'
  }),
  followers: hasMany('user', {
    inverse: 'followees'
  })

  // (...)
});
````
This is two way relation (if you are user C and following user A, then you are the follower of user A). It works pretty good in ember. From the API perspective I want to get both followers and followees as relation from ( JSON API), In mirage I've setup it like this:

````javascript
// mirage/models/user.js
import { Model, hasMany } from 'ember-cli-mirage';

export default Model.extend({
  followees: hasMany('user'),
  followers: hasMany('user')
});

// mirage/serializers/user.js
import { JSONAPISerializer } from 'ember-cli-mirage';

export default JSONAPISerializer.extend({
  relationships: ['followees', 'followers']
});
````
The issue is that it try's to create foreign key `userId` instead of `followeesId`, `followersId`. 

Here are some options:

1. foreign key by default should be reflected from key (and never from target model name - this is how RoR is doing that)
2. add optional params to `hasMany` (ie.: `foreignKey`) so in the end you can overwrite (in RoR it's `as` option)
3. use intermediate model `FollowerFollowee` - still:
  * it will bring me back to my issue with `followees: belongsTo('user'),  followers: belongsTo('user')`
  * it's not recommended way of many to many relationship in ember-data 

I've made very fast patch to get it working (this is not complete solution for many-to-many relation), still I would like to know how to progress with this.  I would really appreciate your feedback, before continuing job.

cs3b/ember-cli-mirage@9aae16ebe442278e64af0eeec53b885df82fc53f

In the end I would like build it together with support for polymorphic relation #433 (and in the end it links somehow to #504)

**Comment by pfgilbert** [comment]
Sorry for the long delay in responding. I haven't had a chance to do anything more than skim through cd3b's comments to see how they relate to my issues, but to answer your questions about how the data relationships are structured on the back end:

*What I've called `top_object` has an array of `vehicle` IDs in its DB record. Some of those vehicle IDs may be in multiple top_objects.
*The `vehicle` object is not an ActiveRecord model and doesn't have any formal associations defined.  Instead, it has a bunch of methods that allow it to gather the correct set of child IDs (`car_ids` and `truck_ids`) at serialization time.

This has gotten a bit difficult to keep my head wrapped around, because of the simplification/obfuscation I've done, so I hope I've described this correctly and that it makes sense to you.

**Comment by samselikoff** [comment]
@cs3b your numbers 1 and 2, especially if that's how RoR does it, I think make sense for us to implement. New issue here: https://github.com/samselikoff/ember-cli-mirage/issues/528

**Comment by samselikoff** [comment]
@pfgilbert thank you for explaining. The `vehicle_ids` stored in your db as an array is the crux of the issue, this is not something I expected of people's backends.

Perhaps we could add an option to mirage's Models to allow the `hasMany` side of the relation to own the keys, though I think this would affect quite a few things. I'll need to go back and think about this when I have some more time.

**Comment by nruth** [comment]
This is a pretty interesting topic! 
There does seem to be a gap in the docs at the moment though.
I'm interested in trying to help out, but have more Ruby/Rails than JS experience. 

TLDR: Does Mirage need an ORM / relational database, or can it just be a JSON / object store?

### docs

> Fixtures are used to seed Mirage's db, and those fixture files should be in the format Mirage expects - regardless of your particular API format. This is because Mirage's ORM expects to find foreign keys on child models. It's independent of the particular serialization format of your API.

http://www.ember-cli-mirage.com/docs/v0.2.0-beta.7/seeding-your-database/#relationships-with-fixtures doesn't really say this at the moment. It shows a positive example, but it's not m2m and doesn't rule out putting an array of child ids on the parent (as you can in a rails AMS serializer for json-api compound documents).

Maybe the docs could explain that mirage fixtures expect the foreign keys to follow the traditional relational model, with foreign keys on the child, and a join model for m2m. Then, ideally it'd say something for people who's data model doesn't have a join table, because it uses arrays instead. Such as, do they need to add a join model for mirage to work with their m2m, or if there's another way.

### Relational model for Mirage

> In Rails (and servers in general), m2m relationships use a join table

If it uses a relational model, but isn't that an implementation detail rather than API design?
Whether that's a good or bad design choice is out of scope here, but some people do it it's worth bearing in mind people do do it. M2M in particular is interesting, as JSON arrays mean you don't need a join table. Much the same as the example with a postgres array mentioned above. 

Even apps that have M2M tables in their RDBMS don't need to expose that in their API if they don't want to? A consumer cares about an exam having many questions, but who cares about some kind of exam-question-usage? They just want to know which questions the exam contains, and perhaps which exams a particular question has been used in.

So I guess an interesting question might be why does Mirage have an ORM? Why does it need to be relational? Isn't it constructing/storing javascript objects?

What we get with Ember-Data is more like ActiveRecord than the database that backs ActiveRecord. Even in Rails, you have HABTM tables where the join records are hidden - it's not always a great design choice (as the relationship often deserves its own model), but it's possible. They're there in the database, but in the object world AR provides you don't see them.

Ember-Data and JSON let us think about objects instead of relational algebra / SQL. In an object model, m2m is an array of ids (or object references) on either side. This is what active-record gives us when it's turned the relational database back-end into objects. With a has-many-through it will give us a list of ids when we ask for it, without us having to think about the join model.

HABTM seem like the real problem with the current approach, as using child IDs means introducing a join model in Mirage that doesn't exist in the real data model / API?

JSON-API supports not having the join model, with compound documents, and Ember-Data seems to figure the rest out, as it's not reliant on a relational model to store relationships.

**Comment by samselikoff** [comment]
I'm definitely open to considering this, though I don't know how hard it would be. If we added HABTM I think that would solve a lot of folks' concerns right now.

To your question about needing an ORM, well it doesn't have to be strict RDBMS style but we do need an ORM (so we can do `post.comments` in the serializer, factory layer etc.).

**Comment by nruth** [comment]
I don't have a better idea at the moment, but is ORM the right name? It makes me think of translating OOP <-> RDBMS?
I'll try to look more into the code, as at the moment I'm speculating.

**Comment by samselikoff** [comment]
Ah. I've been using the term to mean "an object that is hydrated with db data, and is relationship aware". The relationship part is what's important.

**Comment by nruth** [comment]
Yeah that makes sense to me. 

Data-mapper might be an alternative: https://en.wikipedia.org/wiki/Data_mapper_pattern
But it's a bit pedantic/academic. Maybe ORM could be "Object Relationship Manager" :)

**Comment by samselikoff** [comment]
Closing due to inactivity. Hopefully 0-3 series addresses some of the original concerns here.
